README
Test::CPAN::Changes::ReallyStrict

This module is for people who want their Changes file to be 1:1
Identical to how it would be if they'd generated it programmatically
with CPAN::Changes.

This is not for the faint of heart, and will whine about even minor
changes of white-space.

You are also at upstream's mercy as to what a changes file looks like,
and in order to keep this test happy, you'll have to update your whole
changes file if upstream changes how they format things.

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for Test-CPAN-Changes-ReallyStrict

1.000005 2017-03-07T11:01:50Z fb2ef43

1.000004 2017-03-07T11:00:55Z 8ea097f
 - Bugfix: Avoid test failures with -Ddefault_inc_excludes_dot

 [Dependencies::Stats]
 - Dependencies changed since 1.000003, see misc/*.deps* for details
 - develop: +3 ↑2 (suggests: ↑2)

1.000003 2015-07-30T05:03:24Z 14306fe
 [Bugfix]
 - Further narrow the range of versions we do testing on, we missed an important one :(

 [Dependencies::Stats]
 - Dependencies changed since 1.000002, see misc/*.deps* for details
 - test: -1

 [Tests]
 - Centralised version checking logic.
 - Remove use of FindBin

1.000002 2015-07-29T04:58:58Z 2696750
 [Bugfix]
 - Ensure encoding of source and normalised lines are the same.
 - Ensure attempting to load source file in the same manner as CPAN::Changes does w/rt UTF8
 - Guard against test failure if 0.5x radically changes how things format by degrading currently affected tests to TODO

 [Dependencies::Stats]
 - Dependencies changed since 1.000001, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +12 ↑3 -3 (recommends: +1, suggests: ↑2)
 - runtime: +1 ↓1 -1
 - test: +1 (recommends: ↑1 ↓1)

1.000001 2014-08-15T23:33:57Z
 [00 Trivial]
 - CPANDAY!
 - no code changes.
 - tests augmented.
 - metadata updated.
 - whitespace reduced.

 [Dependencies::Stats]
 - Dependencies changed since 1.000000, see misc/*.deps* for details
 - configure: ↓1
 - develop: +1 (suggests: ↑1)
 - test: +1 -3 (recommends: ↑1 -1)

 [Metadata]
 - move to kentnl org.

 [Misc]
 - Improve source side POD to reduce build side whitespace.

 [Tests]
 - update Test::ReportPrereqs

1.000000 2014-07-14T20:00:31Z
 [00 Major]
 - Packaging changes that may break downstreams that work by hand.
 - Internals and stylistic tweaks.

 [01 New Version Scheme]
 - Moving to x.yyyyyy instead of x.y.z ( https://questhub.io/realm/perl/quest/5388d3a9bbd0be2224000129 )
 - This is the primary reason for the 1.0 on the box.

 [02 New Toolkit]
 - This release ships with EUMM instead of MB ( https://questhub.io/realm/perl/quest/5388cc12bbd0be5b2500011f )
 - "latest" Toolkit dependencies softened to "recommends" instead of "requires"

 [Dependencies::Stats]
 - Dependencies changed since 0.2.0, see misc/*.deps* for details
 - build: -1
 - configure: +1 -1 (recommends: +1)
 - develop: +7 -8 (suggests: +2 -1)
 - test: +5 ↓1 (recommends: +4)

0.2.0 2013-12-30T03:31:25Z
 [00 Major]
 - Internals overhauled into a more logical object per-file.
 - Test::MockObject logic eliminated.

 [Dependencies::Noteworthy]
 - Class::Tiny⁺
 - Try::Tiny⁺
 - Test::MockObject¯
 - perl → 5.008

 [Dependencies::Stats]
 - Dependencies changed since 0.1.6, see misc/*.deps* for details
 - develop: ↑1
 - runtime: +3 ↑1
 - test: +1 -1

0.1.6 2013-12-29T06:57:51Z
 [Bugfix]
 - Attempt to fix tests failing on recent CPAN::Changes

 [Dependencies::Noteworthy]
 - CPAN::Changes 0.27 mininum
 - Newer Module::Build and Test::More

 [Dependencies::Stats]
 - Dependencies changed since 0.1.5, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +44 -1 (recommends: -1, suggests: +1 -1)
 - runtime: ↑1 ↓1
 - test: ↑1 -2

 [Test]
 - Use Test::Compile::PerFile

0.1.5 2013-04-09T02:57:17Z
 - Maintenance release for MB 0.4004

 [Dependencies::Noteworthy]
 - Module::Build 0.3601 → 0.4004
 - Test::More 0.96 → 0.98

 [Dependencies::Stats]
 - Dependencies changed since 0.1.4, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +5 (recommends: ↑1, suggests: ↑1)
 - test: ↑1

 [Documentation]
 - Update Copyright year
 - Add README.mkdn

 [Meta]
 - bug tracker to github

 [Packaging]
 - utilise test_requires in Module::Build

0.1.4 2012-02-03T03:51:16Z
 - Maintenance Release

 [Dependencies::Stats]
 - Dependencies changed since 0.1.3, see misc/*.deps* for details
 - develop: (suggests: ↑1)
 - runtime: +3
 - test: -1

 [Internals]
 - name-spaces provide $AUTHORITY
 - $VERSION moved out of BEGIN

 [Packaging]
 - Update LICENSE ( Year, Indentation, Address )
 - Move extra tests to xt/
 - GIT urls moved to https://
 - export x_authority

0.1.3 2011-04-24T13:37:34Z
 - No changes, just this time hopefully the twitter plugin will work because  I actually have LWP::Protocol::https
   installed.

0.1.2 2011-04-24T13:33:38Z
 [Dependencies::Noteworthy]
 - Depend on CPAN::Changes Version 0.17 which contains the surplus whitespace fix.

 [Dependencies::Stats]
 - Dependencies changed since 0.1.1, see misc/*.deps* for details
 - runtime: ↑1

 [Internals]
 - Redundant code that handled <=0.14 versions removed as 0.17 is now a mandatory pre-requisite.

 [Tests]
 - Updated NEXT token extra whitespace test to adapt to changes in CPAN::Changes 0.17

 [Utils::Internals]
 - normalize.pl now uses CPAN::Changes 0.17 instead of hacking the output manually.

0.1.1 2011-04-17T00:38:07Z
 [BugFix]
 - Depend on CPAN::Changes version 0.16 which exposes the regex we consume.

 [Dependencies::Noteworthy]
 - CPAN::Changes now version 0.16 ( 0.160.0 )

 [Dependencies::Stats]
 - Dependencies changed since 0.1.0, see misc/*.deps* for details
 - runtime: ↑1

 [Features]
 - Add utils for migrating/tidying Changes files. zuluize.pl de-dist-zilla-izes it, and normalize runs it through
   CPAN::Changes to tidy it, both reading Changes and writing Changes.out

0.1.0 2011-04-13T20:36:43Z
 - First version.
 - Does Self-Testing for now with a hard-coded example script.


t/lib/mocktest.pm
use strict;
use warnings;
use utf8;

package mocktest;

# ABSTRACT: Mock Test::Builder

use Class::Tiny {
  all_events => sub { [] },
  depth      => sub { 0 },
};

sub ok {
  my ( $self, $code, $reason, @extra ) = @_;
  push @{ $self->all_events }, [ 'ok', $code, $reason, @extra ];
  return $code;
}

sub diag {
  my ( $self, @message ) = @_;
  push @{ $self->all_events }, [ 'diag', @message ];
  return 1;
}

sub note {
  my ( $self, @message ) = @_;
  push @{ $self->all_events }, [ 'note', @message ];
  return 1;
}

sub subtest {
  my ( $self, $desc, $sub ) = @_;

  my $rval;
  push @{ $self->all_events }, [ 'subtest', $desc, 'ENTER', $self->depth ];
  {
    local $self->{depth} = do { $self->{depth} + 1 };
    $rval = $sub->();
  }
  push @{ $self->all_events }, [ 'subtest', $desc, 'EXIT', $self->depth ];
  return $rval;
}

sub ls_events {
  my ($self) = @_;
  return map { join q[, ], @{$_} } @{ $self->all_events };
}

sub num_events {
  my ($self) = @_;
  return scalar @{ $self->all_events };
}

1;



lib/Test/CPAN/Changes/ReallyStrict.pm
use 5.006;
use strict;
use warnings;

package Test::CPAN::Changes::ReallyStrict;

our $VERSION = '1.000005';

#ABSTRACT: Ensure a Changes file looks exactly like it would if it was machine generated.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use CPAN::Changes 0.17;
use Test::Builder;
use Test::CPAN::Changes::ReallyStrict::Object;

my $TEST = Test::Builder->new();

sub import {
  my ( undef, @args ) = @_;

  my $caller = caller;
  {
    ## no critic (ProhibitNoStrict);
    no strict 'refs';
    *{ $caller . '::changes_ok' }      = \&changes_ok;
    *{ $caller . '::changes_file_ok' } = \&changes_file_ok;
  }
  $TEST->exported_to($caller);
  $TEST->plan(@args);
  return 1;
}













sub changes_ok {
  my (@args) = @_;
  return changes_file_ok( undef, @args );
}

# For testing.
sub _real_changes_ok {
  my ( $tester, $state ) = @_;
  return _real_changes_file_ok( $tester, $state );
}















sub changes_file_ok {
  my ( $file, $config ) = @_;
  $file ||= 'Changes';
  $config->{filename} = $file;
  my $changes_obj = Test::CPAN::Changes::ReallyStrict::Object->new(
    {
      testbuilder => $TEST,
      %{$config},
    },
  );
  return $changes_obj->changes_ok;
}

# Factoring design split so testing can inject a test::builder dummy

sub _real_changes_file_ok {
  my ( $tester, $state ) = @_;
  my $changes_obj = Test::CPAN::Changes::ReallyStrict::Object->new(
    {
      testbuilder => $tester,
      %{$state},
    },
  );
  return $changes_obj->changes_ok;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Test::CPAN::Changes::ReallyStrict - Ensure a Changes file looks exactly like it would if it was machine generated.

=head1 VERSION

version 1.000005

=head1 SYNOPSIS

  use Test::More;
  eval 'use Test::CPAN::Changes::ReallyStrict';
  plan skip_all => 'Test::CPAN::Changes::ReallyStrict required for this test' if $@;
  changes_ok();
  done_testing();

=head1 DESCRIPTION

This module is for people who want their Changes file to be 1:1 Identical to how it would be
if they'd generated it programmatically with CPAN::Changes.

This is not for the faint of heart, and will whine about even minor changes of white-space.

You are also at upstream's mercy as to what a changes file looks like, and in order to keep this test
happy, you'll have to update your whole changes file if upstream changes how they format things.

=head1 EXPORTED FUNCTIONS

=head2 changes_ok

  changes_ok();

  changes_ok({
    delete_empty_groups => 1,
    keep_comparing => 1,
    next_style => 'dzil'
  });

=head2 changes_file_ok

  changes_file_ok();

  changes_file_ok('ChangeLog');

  changes_ok('ChangeLog', {
    delete_empty_groups => 1,
    keep_comparing => 1,
    next_style => 'dzil'
  });

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Test/CPAN/Changes/ReallyStrict/Object.pm
use 5.006;
use strict;
use warnings;

package Test::CPAN::Changes::ReallyStrict::Object;

our $VERSION = '1.000005';

# ABSTRACT: Object Oriented Guts to ::ReallyStrict

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Test::Builder;
use Encode qw( decode FB_CROAK LEAVE_SRC );
use Try::Tiny qw( try catch );

my $TEST       = Test::Builder->new();
my $version_re = '^[._\-[:alnum:]]+$';    # "Looks like" a version

use Class::Tiny {
  testbuilder => sub { $TEST },
  filename    => sub { 'Changes' },
  next_token  => sub {
    return unless defined $_[0]->next_style;
    return qr/[{][{]\$NEXT[}][}]/msx if 'dzil' eq $_[0]->next_style;
    return;
  },
  next_style => sub { undef },
  changes    => sub {
    my ($self) = @_;
    require CPAN::Changes;
    my @extra;
    push @extra, ( next_token => $self->next_token ) if defined $self->next_token;
    return CPAN::Changes->load( $self->filename, @extra );
  },
  normalised_lines => sub {
    my ($self) = @_;
    if ( $self->delete_empty_groups ) {
      $self->changes->delete_empty_groups;
    }
    my $string = $self->changes->serialize;
    return [ split /\n/msx, $string ];
  },
  source_lines => sub {
    my ($self) = @_;
    my $fh;
    ## no critic (ProhibitPunctuationVars)

    if ( not open $fh, '<:raw', $self->filename ) {
      $self->testbuilder->ok( 0, $self->filename . ' failed to open' );
      $self->testbuilder->diag( 'Error ' . $! );
      return;
    }
    my $str = do {
      local $/ = undef;
      scalar <$fh>;
    };
    close $fh or $self->testbuilder->diag( 'Warning: Error Closing ' . $self->filename );
    ## no critic (RequireCheckingReturnValueOfEval, ProhibitBitwiseOperators)
    eval { $str = decode( 'UTF-8', $str, FB_CROAK | LEAVE_SRC ); };
    return [ split /\n/msx, $str ];
  },
  delete_empty_groups => sub { },
  keep_comparing      => sub { },
};





sub changes_ok {
  my ( $self, ) = @_;
  my $exi;
  $self->testbuilder->subtest(
    'changes_ok' => sub {
      return unless $self->loads_ok;
      return unless $self->has_releases;
      return unless $self->valid_releases;
      return unless $self->compare_lines;

      #$self->testbuilder->ok(1, 'All Subtests for ' . $self->filename . ' done' );
      $exi = 1;
    },
  );
  return unless $exi;
  return 1;
}









sub loads_ok {
  my ($self) = @_;
  my ( $error, $success );
  try {
    $self->changes();
    $success = 1;
  }
  catch {
    undef $success;
    $error = $_;
  };
  if ( not $error and $success ) {
    $self->testbuilder->ok( 1, $self->filename . ' is loadable' );
    return 1;
  }
  $self->testbuilder->ok( 0, $self->filename . ' is loadable' );
  $self->testbuilder->diag($error);
  return;
}









sub has_releases {
  my ($self)     = @_;
  my (@releases) = $self->changes->releases;
  if (@releases) {
    $self->testbuilder->ok( 1, $self->filename . ' contains at least one release' );
    return 1;
  }
  $self->testbuilder->ok( 0, $self->filename . ' does not contain any release' );
  return;
}









sub valid_release_date {
  my ( $self, $release, $release_id ) = @_;
  if ( not defined $release->date and defined $self->next_token ) {
    $self->testbuilder->ok( 1, "release $release_id has valid date (none|next_token)" );
    return 1;
  }
  if ( $release->date =~ m/\A${CPAN::Changes::W3CDTF_REGEX}\s*\z/msx ) {
    $self->testbuilder->ok( 1, "release $release_id has valid date (regexp match)" );
    return 1;
  }
  $self->testbuilder->ok( 0, "release $release_id has an invalid release date" );
  $self->testbuilder->diag( '  ERR:' . $release->date );
  return;
}









sub valid_release_version {
  my ( $self, $release, $release_id ) = @_;
  if ( not defined $release->version and defined $self->next_token ) {
    $self->testbuilder->ok( 1, "release $release_id has valid version (none|next_token)" );
    return 1;
  }
  if ( defined $self->next_token and $release->version =~ $self->next_token ) {
    $self->testbuilder->ok( 1, "release $release_id has valid version (regexp match on next_token)" );
    return 1;
  }
  if ( $release->version =~ m/$version_re/msx ) {
    $self->testbuilder->ok( 1, "release $release_id has valid version (regexp match version re)" );
    return 1;
  }
  $self->testbuilder->ok( 0, "release $release_id has valid version." );
  $self->testbuilder->diag( '  ERR:' . $release->version );
  return;
}









sub valid_releases {
  my ($self) = @_;
  my $top_exit = 1;

  $self->testbuilder->subtest(
    'valid releases' => sub {
      my (@releases) = $self->changes->releases;
      for my $id ( 0 .. $#releases ) {
        my ($release) = $releases[$id];
        my $sub_exit;
        $self->testbuilder->subtest(
          'valid release: ' . $id => sub {
            return unless $self->valid_release_date( $release, $id );
            return unless $self->valid_release_version( $release, $id );
            $sub_exit = 1;
          },
        );
        undef $top_exit unless $sub_exit;
      }
    },
  );
  return 1 if $top_exit;
  return;
}









sub compare_line {
  my ( $self, $source, $normalised, $line_number, $failed_before ) = @_;
  if ( not defined $source and not defined $normalised ) {
    $self->testbuilder->ok( 1, "source($line_number) == normalised($line_number) : undef vs undef" );
    return 1;
  }
  if ( defined $source and not defined $normalised ) {
    $self->testbuilder->ok( 0, "source($line_number) != normalised($line_number) : defined vs undef" );
    return;
  }
  if ( not defined $source and defined $normalised ) {
    $self->testbuilder->ok( 0, "source($line_number) != normalised($line_number) : undef vs defined" );
    return;
  }
  if ( $] > 5.008 ) {
    ## no critic (ProhibitCallsToUnexportedSubs)
    if ( $ENV{AUTHOR_TESTING} ) {
      my (@utf8ness) = map { utf8::is_utf8($_) } $source, $normalised;
      if ( $utf8ness[0] != $utf8ness[1] ) {
        $self->testbuilder->diag( sprintf 'utf8ness differs: source=%s normalised=%s', @utf8ness );
      }
    }
    utf8::encode($source)     if utf8::is_utf8($source);
    utf8::encode($normalised) if utf8::is_utf8($normalised);
  }
  if ( $source eq $normalised ) {
    $self->testbuilder->ok( 1, "source($line_number) == normalised($line_number) : val eq val" );
    return 1;
  }
  if ( not $failed_before ) {
    $self->testbuilder->ok( 0, "Lines differ at $line_number" );
  }
  $self->testbuilder->diag( sprintf q{[%s] Expected: >%s<}, $line_number, $normalised );
  $self->testbuilder->diag( sprintf q{[%s] Got     : >%s<}, $line_number, $source );
  return;

}









sub compare_lines {
  my ($self) = @_;

  my (@source)     = @{ $self->source_lines };
  my (@normalised) = @{ $self->normalised_lines };

  my $all_lines_passed = 1;

  $self->testbuilder->subtest(
    'compare lines source vs normalised' => sub {
      $self->testbuilder->note( sprintf q[Source: %s, Normalised: %s], $#source, $#normalised );
      my $failed_already;
      for ( 0 .. $#source ) {
        my $line_passed = $self->compare_line( $source[$_], $normalised[$_], $_, $failed_already );
        if ( not $line_passed ) {
          $failed_already = 1;
          undef $all_lines_passed;
          if ( not $self->keep_comparing ) {
            last;
          }
        }
      }
    },
  );
  return 1 if $all_lines_passed;
  return;
}
1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Test::CPAN::Changes::ReallyStrict::Object - Object Oriented Guts to ::ReallyStrict

=head1 VERSION

version 1.000005

=head1 METHODS

=head2 C<changes_ok>

=head2 C<loads_ok>

    if ( $self->loads_ok() ) {

    }

=head2 C<has_releases>

    if( $self->has_releases() ){

    }

=head2 C<valid_release_date>

    if ( $self->valid_release_date( $release, $release_id ) ) {

    }

=head2 C<valid_release_version>

    if ( $self->valid_release_version( $release, $release_id ) ) {

    }

=head2 C<valid_releases>

    if ( $self->valid_releases() ) {

    }

=head2 C<compare_line>

    if ( $self->compare_line( $source_line, $normalised_line, $line_number, $failed_before ) ) {

    }

=head2 C<compare_lines>

    if ( $self->compare_lines ) {

    }

=head1 ATTRIBUTES

=head2 C<testbuilder>

Plumbing: This is where test builder calls get made.

=head2 C<filename>

The name/path of the changes file.

B<Default>: C<Changes>

=head2 C<next_token>

The regular expression to use for C<next_token>

Defaults to C<undef>, or C<{{$NEXT}}> if C<next_style> C<eq> C<dzil>

=head2 C<next_style>

The C<next_token> style.

Defaults to C<undef>

=head2 C<changes>

B<Lazy>: A C<CPAN::Changes> object read from C<filename>

=head2 C<normalised_lines>

B<Lazy>: Lines from serializing C<changes>

=head2 C<source_lines>

B<Lazy>: Lines from C<filename>

=head2 C<delete_empty_groups>

B<Default>: C<undef>

Whether to delete empty groups while serializing.

=head2 C<keep_comparing>

B<Default>: C<undef>

Whether to continue comparing lines after a miss-match.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Ensure a Changes file looks exactly like it would if it was machine generated.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Test-CPAN-Changes-ReallyStrict",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Test::CPAN::Changes::ReallyStrict",
  "PREREQ_PM" => {
    "CPAN::Changes" => "0.27",
    "Class::Tiny" => 0,
    "Encode" => 0,
    "Test::Builder" => 0,
    "Try::Tiny" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => "0.96",
    "Text::Wrap" => 0,
    "lib" => 0
  },
  "VERSION" => "1.000005",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "CPAN::Changes" => "0.27",
  "Class::Tiny" => 0,
  "Encode" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Test::Builder" => 0,
  "Test::More" => "0.96",
  "Text::Wrap" => 0,
  "Try::Tiny" => 0,
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


util/zuluize.pl
#!/usr/bin/perl

# Replace DZIL style date-time with W3CDTF style while keeping the time right.

use strict;
use warnings;

use DateTime::Format::Strptime;

open my $ifh, '<', 'Changes' or die "Can't open Changes, $!";
open my $ofh, '>', 'Changes.out' or die "Can't open Changes.out for write, $!";

while (defined(my $line = <$ifh>)) {
  if ( $line =~ qr{(^[\d.]+\s+)(\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2})\s+([^\s]+)} ) {
    my $prelude = $1;
    my $date    = $2;
    my $timezone = $3;
    print "TS: $2 , TZ: $3\n";
    my $strp = DateTime::Format::Strptime->new(
      pattern   => '%Y-%m-%d %H:%M:%S',
      time_zone => $timezone,
    );
    $line =  join '', ( $prelude, $strp->parse_datetime($date)->set_time_zone('UTC'), "Z\n" );
  } elsif ( $line =~ /^\s*\d/ ){
    my $lineclone = $line;
    chomp $lineclone;
    print "Datestamp line $lineclone skipped, probably already ok\n";
  }
  print { $ofh } $line;
}




maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (
  qw(
    programmatically
    )
);
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

#$bundle->add_or_append_policy_field(
#  'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );

$bundle->remove_policy('ErrorHandling::RequireUseOfExceptions');

#$bundle->remove_policy('CodeLayout::RequireUseUTF8');

#$bundle->remove_policy('ErrorHandling::RequireCarping');
#$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  my $rcfile = path('./perlcritic.rc')->openw_utf8;
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  my $target = path('./misc');
  $target->mkpath if not $target->is_dir;

  my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.019.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4206"
  },
  "configure_requires" => {
    "Module::Build" => "0.4206"
  },
  "dist_abstract" => "Ensure a Changes file looks exactly like it would if it was machine generated.",
  "dist_author" => [
    "Kent Fredric <kentnl\@cpan.org>"
  ],
  "dist_name" => "Test-CPAN-Changes-ReallyStrict",
  "dist_version" => "1.000000",
  "license" => "perl",
  "module_name" => "Test::CPAN::Changes::ReallyStrict",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "CPAN::Changes" => "0.27",
    "Class::Tiny" => 0,
    "Test::Builder" => 0,
    "Try::Tiny" => 0,
    "perl" => "5.008",
    "strict" => 0,
    "utf8" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec::Functions" => 0,
    "FindBin" => 0,
    "List::Util" => 0,
    "Test::More" => "1.001003",
    "lib" => 0,
    "version" => 0
  }
);


my %fallback_build_requires = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec::Functions" => 0,
  "FindBin" => 0,
  "List::Util" => 0,
  "Module::Build" => "0.4206",
  "Test::More" => "1.001003",
  "lib" => 0,
  "version" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
